From 572f747ebad0a7d4caaedc4d68ac83e27ab2de0e Mon Sep 17 00:00:00 2001
From: Patrick Gaskin <patrick@pgaskin.net>
Date: Thu, 31 Dec 2020 05:03:09 -0500
Subject: [PATCH] win32: Fix duplicate definitions in arpa-inet.h, arpa-inet.c,
 and poll.h

When _WIN32_WINNT >= 0x6000 (Vista), ws2tcpip.h provides inet_ntop and
inet_pton, and winsock2.h provides the equivalent of poll.h.
---
 src/pulsecore/arpa-inet.c | 4 ++++
 src/pulsecore/arpa-inet.h | 6 +++++-
 src/pulsecore/poll.h      | 2 ++
 3 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/src/pulsecore/arpa-inet.c b/src/pulsecore/arpa-inet.c
index afea3971a..b177dcc5f 100644
--- a/src/pulsecore/arpa-inet.c
+++ b/src/pulsecore/arpa-inet.c
@@ -31,6 +31,8 @@
 
 #include "arpa-inet.h"
 
+#if NTDDI_VERSION < NTDDI_VISTA
+
 const char *inet_ntop(int af, const void *src, char *dst, socklen_t cnt) {
     struct in_addr *in = (struct in_addr*)src;
 #ifdef HAVE_IPV6
@@ -104,3 +106,5 @@ int inet_pton(int af, const char *src, void *dst) {
 }
 
 #endif
+
+#endif
diff --git a/src/pulsecore/arpa-inet.h b/src/pulsecore/arpa-inet.h
index d940f704f..775c2269d 100644
--- a/src/pulsecore/arpa-inet.h
+++ b/src/pulsecore/arpa-inet.h
@@ -8,10 +8,12 @@
 #elif defined(OS_IS_WIN32)
 
 /* On Windows winsock2.h (here included via pulsecore/socket.h) provides most of the functionality of arpa/inet.h, except for
- * the inet_ntop and inet_pton functions, which are implemented here. */
+ * the inet_ntop and inet_pton functions, which are implemented here on versions earlier than Vista. */
 
 #include <pulsecore/socket.h>
 
+#if NTDDI_VERSION < NTDDI_VISTA
+
 const char *inet_ntop(int af, const void *src, char *dst, socklen_t cnt);
 
 int inet_pton(int af, const char *src, void *dst);
@@ -19,3 +21,5 @@ int inet_pton(int af, const char *src, void *dst);
 #endif
 
 #endif
+
+#endif
diff --git a/src/pulsecore/poll.h b/src/pulsecore/poll.h
index 4af1b9946..af34506c5 100644
--- a/src/pulsecore/poll.h
+++ b/src/pulsecore/poll.h
@@ -24,6 +24,8 @@
 
 #if defined(HAVE_POLL_H)
 #include <poll.h>
+#elif OS_IS_WIN32 && HAVE_WINSOCK2_H && NTDDI_VERSION >= NTDDI_VISTA
+#include <winsock2.h>
 #else
 
 /* Event types that can be polled for.  These bits may be set in `events'
-- 
2.29.2

